test/IntegrationTests/MSTest.Acceptance.IntegrationTests/LeakTests.cs

```diff
@@ -0,0 +1,132 @@
+        {
+            yield return (ProjectName, ProjectName,
+                SourceCode
+                .PatchTargetFrameworks(TargetFrameworks.All)
+                .PatchCodeWithReplace("$MSTestVersion$", MSTestVersion));
+        }
+
+        private const string SourceCode = """
+#file LeakTests.csproj
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <OutputType>Exe</OutputType>
+    <EnableMSTestRunner>true</EnableMSTestRunner>
+    <TargetFrameworks>$TargetFrameworks$</TargetFrameworks>
+    <LangVersion>preview</LangVersion>
+
+    <!--
+        This property is not required by users and is only set to simplify our testing infrastructure. When testing out in local or ci,
+        we end up with a -dev or -ci version which will lose resolution over -preview dependency of code coverage. Because we want to
+        ensure we are testing with locally built version, we force adding the platform dependency.
+    -->
+    <EnableMicrosoftTestingPlatform>true</EnableMicrosoftTestingPlatform>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="MSTest" Version="$MSTestVersion$" />
+  </ItemGroup>
+
+</Project>
+
+#file UnitTest1.cs
+using System;
+using System.Collections.Generic;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+
+[TestClass]
+public class TestClass
+{
+    private static ConcurrentBag<WeakReference<TestContext>> _testContexts = new();
+
+    [AssemblyInitialize]
+    public static void AssemblyInitialize(TestContext context)
+        => _testContexts.Add(new WeakReference<TestContext>(context));
+
+    [ClassInitialize]
+    public static void ClassInitialize(TestContext context)
+        => _testContexts.Add(new WeakReference<TestContext>(context));
+
+    public TestContext TestContext { get; set; }
+
+    [TestMethod]
+    [DynamicData(nameof(Data))]
+    public void Test3(int a)
+        => _testContexts.Add(new WeakReference<TestContext>(TestContext));
+
+    [ClassCleanup]
+    public static void ClassCleanup(TestContext testContext)
+        => _testContexts.Add(new WeakReference<TestContext>(testContext));
+
+    [AssemblyCleanup]
+    public static void AssemblyCleanup()
+    {
+        for (int i = 0; i < 3; i++)
+        {
+            GC.Collect();
+            GC.WaitForPendingFinalizers();
+        }
+
+        // Assembly init, class init, 100 tests, and class cleanup. (total 103).
+        Assert.AreEqual(103, _testContexts.Count);
+
+        var alive = 0;
+        foreach (var weakReference in _testContexts)
+        {
+            if (weakReference.TryGetTarget(out _))
+            {
+                alive++;
+            }
+        }
+
+        // AssemblyCleanup is executed along with the last test.
+        // So, we are still holding 2 references to the TestContext. The one for the execution of last test, as well as the one for ClassCleanup.
+        // Holding into these two references is okay.
+        Assert.AreEqual(2, alive);
```

Comment:
Did a quick test with that code in Playground.

On `main`, all the 103 instances are alive. With the fix, only 2 are alive which is good.